### PR TITLE
Add support for Custom Policies for Guest Configuration cmdlets

### PR DESCRIPTION
**Pull request by aashish1987**
## Description

Support reports retrieval for custom policies as well.
Add version information for the report.

## Checklist

- [X] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md)
- [X] The title of the PR is clear and informative
- [X] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md#updating-the-change-log)
- [X] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [X] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (Microsoft internal partners only)
    - [ ] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/help-generation.md)

### CODE REVIEW

**Comment by maddieclayton** [comment]
@aashish1987 Please either link your cmdlet review or fill out a new one here: https://github.com/Azure/azure-powershell-cmdlet-review-pr

**Comment by aashish1987** [comment]
Cmd-let review links -
https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/187
https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/118

File: src/GuestConfiguration/GuestConfiguration/Common/GuestConfigurationCmdletBase.cs [hunk 1]
```diff
@@ -310,7 +322,16 @@ private IEnumerable<PolicyData> GetPolicyStatusesHelper(PolicySetDefinition[] po
                 foreach (var policyDefinitionIdInInitiative in policyDefinitionIdsInInitiative)
                 {
                     var _initiativeName = GetInitiativeNameFromId(policyDefinitionIdInInitiative);
-                    var policyDef = PolicyClient.PolicyDefinitions.GetBuiltIn(_initiativeName);
+                    PolicyDefinition policyDef;
+                    try
+                    {
+                        policyDef = PolicyClient.PolicyDefinitions.GetBuiltIn(_initiativeName);
```
Comment (vivlingaiah) [inline src/GuestConfiguration/GuestConfiguration/Common/GuestConfigurationCmdletBase.cs]: how do you distinguish between service error vs builtin/custom policy ?

File: src/GuestConfiguration/GuestConfiguration/Common/GuestConfigurationCmdletBase.cs [hunk 2]
```diff
@@ -271,13 +278,18 @@ private string GetPolicySetCategory(PolicySetDefinition initiativeDefinition)
...
+                return category;
+            }
+
+            return string.Empty;
         }
 
         private IEnumerable<PolicySetDefinition> GetAllGuestConfigPolicySetDefinitions()
         {
-            var policySetDefinitions = PolicyClient.PolicySetDefinitions.ListBuiltIn();
+            var policySetDefinitions = PolicyClient.PolicySetDefinitions.List();
```
Comment (vivlingaiah) [inline src/GuestConfiguration/GuestConfiguration/Common/GuestConfigurationCmdletBase.cs]: List returns custom policies and excludes builtin policies. can you make sure after cmdlet works on builtin initiatives as well with your changes ?

File: src/GuestConfiguration/GuestConfiguration/Common/GuestConfigurationCmdletBase.cs [hunk 3]
```diff
@@ -271,13 +278,18 @@ private string GetPolicySetCategory(PolicySetDefinition initiativeDefinition)
         {
             var categoryMetadata = JObject.Parse(initiativeDefinition.Metadata.ToString());
             var categoryMetadataDictionary = categoryMetadata.ToObject<Dictionary<string, object>>();
-            var category = categoryMetadataDictionary["category"].ToString();
-            return category;
+            if (categoryMetadataDictionary.ContainsKey("category"))
```
Comment (vivlingaiah) [inline src/GuestConfiguration/GuestConfiguration/Common/GuestConfigurationCmdletBase.cs]: use constant for "category"

Comment (aashish1987) [reply src/GuestConfiguration/GuestConfiguration/Common/GuestConfigurationCmdletBase.cs]: Wont fix for now.
String literals have been used everywhere in the file.
I can take up a new task to fix the string constants everywhere in the cmdlets.

---
In reply to: [258726975](https://github.com/Azure/azure-powershell/pull/8574#discussion_r258726975) [](ancestors = 258726975)

File: src/GuestConfiguration/GuestConfiguration/Common/GuestConfigurationCmdletBase.cs (hunk 1 above)
Comment (aashish1987) [reply src/GuestConfiguration/GuestConfiguration/Common/GuestConfigurationCmdletBase.cs]: catching specific exception and adding when clause corresponding to not found status code.

---
In reply to: [258725693](https://github.com/Azure/azure-powershell/pull/8574#discussion_r258725693) [](ancestors = 258725693)

Comment (vivlingaiah) [reply src/GuestConfiguration/GuestConfiguration/Common/GuestConfigurationCmdletBase.cs]: have you tested this works for both custom and builtin policies ?

**Comment by maddieclayton** [comment]
@aashish1987 Please fix merge conflicts

**Comment by aashish1987** [comment]
@maddieclayton - I have finished merging with latest branch.

File: src/GuestConfiguration/GuestConfiguration.Test/ScenarioTests/GetVmGuestPolicyStatusHistoryTests.ps1 [hunk 4]
```diff
@@ -41,6 +55,21 @@ function Get-AzVMGuestPolicyStatusHistory-InitiativeIdScope
...
 
+<#
+.SYNOPSIS
+Get guest configuration policy status history by Initiative id scope
+#>
+function Get-AzVMGuestPolicyStatusHistory-InitiativeIdScope_Custom
+{
+	$rgName = "amits-test"
+	$vmName = "amits-winser6"
+	$initiativeId = "/subscriptions/b5e4748c-f69a-467c-8749-e2f9c8cd3db0/providers/Microsoft.Authorization/policySetDefinitions/8174929b-2752-46af-9167-89d2d6780439"
```
Comment (cormacpayne) [inline src/GuestConfiguration/GuestConfiguration.Test/ScenarioTests/GetVmGuestPolicyStatusHistoryTests.ps1]: @aashish1987 in order to allow other users to record these tests if needed, we shouldn't hard-code resource ids, as it's not guaranteed the user recording the test has access to this subscription or definition. Tests should properly construct all necessary resources and objects that will be used during the test, and also delete them at the end of every test.

File: src/GuestConfiguration/GuestConfiguration/Common/GuestConfigurationCmdletBase.cs [hunk 5]
```diff
@@ -271,13 +278,18 @@ private string GetPolicySetCategory(PolicySetDefinition initiativeDefinition)
...
+            {
+                var category = categoryMetadataDictionary["category"].ToString();
```
Comment (cormacpayne) [inline src/GuestConfiguration/GuestConfiguration/Common/GuestConfigurationCmdletBase.cs]: @aashish1987 minor: just for simplification:

```cs
return categoryMetadataDictionary["category"].ToString();
```

File: tools/ScenarioTest.ResourceManager/ScenarioTest.ResourceManager.csproj [hunk 6]
```diff
@@ -9,6 +9,7 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <PackageReference Include="Microsoft.Azure.Management.GuestConfiguration" Version="1.1.0" />
```
Comment (cormacpayne) [inline tools/ScenarioTest.ResourceManager/ScenarioTest.ResourceManager.csproj]: @aashish1987 please remove this, this shouldn't be needed

File: src/GuestConfiguration/GuestConfiguration/Az.GuestConfiguration.psd1 [hunk 7]
```diff
@@ -12,7 +12,7 @@
 # RootModule = ''
 
 # Version number of this module.
-ModuleVersion = '0.10.2'
+ModuleVersion = '0.10.3'
```
Comment (cormacpayne) [inline src/GuestConfiguration/GuestConfiguration/Az.GuestConfiguration.psd1]: @aashish1987 you can revert this change since we'll do the versioning before the module is released

File: src/GuestConfiguration/GuestConfiguration/ChangeLog.md [hunk 8]
```diff
@@ -19,6 +19,10 @@
 -->
 ## Upcoming Release
 
+## Version 0.10.3
```
Comment (cormacpayne) [inline src/GuestConfiguration/GuestConfiguration/ChangeLog.md]: @aashish1987 same comment here, you can remove this header and put the below two lines under `## Upcoming Release` since we'll do versioning before the module is released

File: src/GuestConfiguration/GuestConfiguration.Test/ScenarioTests/GetVmGuestPolicyStatusHistoryTests.ps1 (hunk 4 above)
Comment (aashish1987) [reply src/GuestConfiguration/GuestConfiguration.Test/ScenarioTests/GetVmGuestPolicyStatusHistoryTests.ps1]: @corm@microsoft.com - For Guest Configuration, i did follow what was being used in past.
I totally agree we should not hard code resources as it might result in access issues for others to record the tests.
For now I will mark this as wont fix. 
I will take this change up in a new PR.

---
In reply to: [261284886](https://github.com/Azure/azure-powershell/pull/8574#discussion_r261284886) [](ancestors = 261284886)
